File: lintcode/python/0453_flatten_binary_tree_to_linked_list.py
#!/usr/bin/python -t

# bfs non-recursive way

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def flatten(self, root):
        # write your code here
        if root == None:
            return
        
        s = [root]
        
        while len(s) != 0:
            node = s.pop()
            
            if node.right:
                s.append(node.right)
            if node.left:
                s.append(node.left)
                
            node.left = None
            if len(s) == 0:
                node.right = None
            else:
                node.right = s[-1]
                
        return


# dfs traversal solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def __init__(self):
        self.lastnode = None
        
    def flatten(self, root):
        # write your code here
        if root == None:
            return
        
        if self.lastnode != None:
            self.lastnode.right = root
            self.lastnode.left = None
            
        self.lastnode = root
        rightnode = root.right
        self.flatten(root.left)
        self.flatten(rightnode)
        
        return
    

# pure divid and conquer solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def flatten(self, root):
        # write your code here
        if root == None:
            return
        
        self.helper(root)
        
        return
    
    def helper(self, node):
        if node == None:
            return
        
        l_last = self.helper(node.left)
        r_last = self.helper(node.right)
        
        if l_last != None:
            l_last.right = node.right
            node.right = node.left
            node.left = None
            
        if r_last != None:
            return r_last
            
        if l_last != None:
            return l_last
            
        return node
        

File: lintcode/golang/0453_flatten_binary_tree_to_linked_list.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: a TreeNode, the root of the binary tree
 * @return: nothing
 */
func flatten (root **TreeNode)  {
    // write your code here
    if *root == nil {
        return
    }
    
    _ = helper((*root))

    return
}

func helper(node *TreeNode) *TreeNode {
    if node == nil {
        return nil
    }
    
    leftLast := helper(node.Left)
    rightLast := helper(node.Right)
    
    if leftLast != nil {
        leftLast.Right = node.Right
        node.Right = node.Left
        node.Left = nil

    }
    
    if rightLast != nil {
        return rightLast
    }
    if leftLast != nil {
        return leftLast
    }
    
    return node
}
